provider/pkg/provider/local/commandOutputs.go

```diff
@@ -46,18 +46,13 @@ func (c *CommandOutputs) run(ctx p.Context, command string) (string, string, err
 	}
 	args = append(args, command)
 
-	stdoutr, stdoutw, err := os.Pipe()
-	if err != nil {
-		return "", "", err
-	}
-	stderrr, stderrw, err := os.Pipe()
-	if err != nil {
-		return "", "", err
-	}
+	var err error
+	var stdoutbuf, stderrbuf, stdouterrbuf bytes.Buffer
+	r, w := io.Pipe()
 
 	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
-	cmd.Stdout = stdoutw
-	cmd.Stderr = stderrw
+	cmd.Stdout = io.MultiWriter(&stdoutbuf, &stdouterrbuf, w)
+	cmd.Stderr = io.MultiWriter(&stderrbuf, &stdouterrbuf, w)
```

Comment:
I don't know if this is a concern here, but this may cause a data race in writing to stdouterrbuf because os/exec will spawn separate goroutines to read from the stdout and stderr pipes and write to the provided writers.

os/exec has a special case for when Stdout and Stderr are the same writer, but that won't apply here.

```
	// If Stdout and Stderr are the same writer, and have a type that can
	// be compared with ==, at most one goroutine at a time will call Write.
```

(https://pkg.go.dev/os/exec#Cmd.Stdout)

If that's a concern here, we may want to wrap the bytes.Buffer's Write method with a mutex. Something simple like the following would suffice.

```go
type lockedWriter struct {
  w  io.Writer
  mu sync.Mutex
}

func (w *lockedWriter) Write(bs []byte) (int, error) {
  w.mu.Lock()
  n, err := w.w.Write(bs)
  w.mu.Unlock()
}
```

(This isn't an issue for `io.Pipe` so that `w` shouldn't need this.)

Reply:
> os/exec has a special case for when Stdout and Stderr are the same writer, but that won't apply here.

Great callout!  Indeed there's an even deeper issue due to this, that we can't guarantee we present a correctly ordered view of the combined stdout+stderr unless we pass them the same writer, and if we do that, we can't break them apart.  So in Go today, it appears there is no way to get both (1) the ability to see the stdout and stderr streams independently and (2) a correctly ordered single stream of stdout+stderr in the order the process emitted them.  Now of course (2) is probably not even 100% well defined at some even lower level, but we can lose arbitrary correctness on this just at the Go level.  After some disbelief around this, I finally found a confirmation that this was a limitation in Go os/exec, but can't find the reference now.

> If that's a concern here, we may want to wrap the bytes.Buffer's Write method with a mutex

Great point on the concurrency concern on the bytes.Buffer, will add the mutex to guard against this.